compiler/rustc_infer/src/traits/mod.rs

```diff
@@ -105,6 +109,65 @@ pub enum FulfillmentErrorCode<'tcx> {
     CodeAmbiguity,
 }
 
+pub struct ObligationsDedup<'a, 'tcx, T> {
+    obligations: &'a mut Vec<Obligation<'tcx, T>>,
+}
+
+impl<'a, 'tcx, T: 'tcx> ObligationsDedup<'a, 'tcx, T>
+where
+    T: Clone + Hash + Eq,
+{
+    pub fn from_vec(vec: &'a mut Vec<Obligation<'tcx, T>>) -> Self {
+        ObligationsDedup { obligations: vec }
+    }
+
+    pub fn extend<'b>(&mut self, iter: impl ExactSizeIterator<Item = Cow<'b, Obligation<'tcx, T>>>)
+    where
+        'tcx: 'b,
+    {
+        // obligation tracing has shown that initial batches added to an empty vec do not
+        // contain any duplicates, so there's no need to attempt deduplication
+        if self.obligations.is_empty() {
+            *self.obligations = iter.into_iter().map(Cow::into_owned).collect();
+            return;
+        }
+
+        let initial_size = self.obligations.len();
+        let current_capacity = self.obligations.capacity();
+        let iter = iter.into_iter();
+        let expected_new = iter.len();
+        let combined_size = initial_size + expected_new;
+
+        if combined_size <= 16 || combined_size <= current_capacity {
+            // small case/not crossing a power of two. don't bother with dedup
+            self.obligations.extend(iter.map(Cow::into_owned));
+        } else {
+            // crossing power of two threshold. this would incur a vec growth anyway if we didn't do
+            // anything. piggyback a dedup on that
+            let mut seen = FxHashMap::default();
+            seen.reserve(initial_size);
```

Comment:
Isn't this the same as the following?

```suggestion
            let mut seen = FxHashMap::with_capacity(initial_size);
```